File: web/js/views/strategy.js
/* Strategy: multi-leg payoff, mark-to-market P&L at intermediate dates, and
   aggregate position Greeks. */

import { el, clear, buildTable, fmtSig, fmtSigned, fmtMoney, fmtPct, seriesColor, token } from '../util.js';
import { api } from '../api.js';
import { state, signature } from '../state.js';
import { LineChart, withTableToggle } from '../chart.js';

const PRESETS = [
  ['long_call', 'Long call'],
  ['long_put', 'Long put'],
  ['covered_call', 'Covered call'],
  ['bull_call_spread', 'Bull call spread'],
  ['bear_put_spread', 'Bear put spread'],
  ['straddle', 'Long straddle'],
  ['short_strangle', 'Short strangle'],
  ['iron_condor', 'Iron condor'],
  ['butterfly', 'Call butterfly'],
  ['calendar_ratio', 'Call ratio 1x2'],
];

export class StrategyView {
  constructor(host, ctx) {
    this.host = host;
    this.ctx = ctx;
    this.dirty = true;
    this.legs = null;
    this.preset = 'iron_condor';
    this._build();
  }

  _build() {
    clear(this.host);
    this.notice = el('div');

    this.presetChips = el('div', { class: 'chip-row' },
      ...PRESETS.map(([k, l]) => el('button', {
        class: 'chip', type: 'button',
        onclick: () => { this.preset = k; this.legs = null; this.refresh().catch((e) => this.ctx.reportError(e, 'strategy')); },
      }, l)));

    this.legsBox = el('div');
    this.addBtn = el('button', {
      class: 'btn btn-ghost btn-sm', type: 'button',
      onclick: () => {
        this.legs = [...(this.legs || []), { kind: 'call', quantity: 1, strike: Math.round(state.S), sigma: state.sigma }];
        this._renderLegs();
        this._recalc();
      },
    }, '+ Add leg');

    const builderCard = el('div', { class: 'card' },
      el('div', { class: 'card-head' },
        el('div', {},
          el('h3', {}, 'Position'),
          el('p', {}, 'Quantities are signed: negative is short. Each leg carries its own implied '
            + 'volatility, so a skewed structure can be marked with the skew it was actually traded at.')),
        el('div', { class: 'card-tools' }, this.addBtn)),
      el('div', { class: 'panel-title' }, 'Presets'),
      this.presetChips,
      this.legsBox);

    this.payoffHost = el('div');
    this.payoffLegend = el('div');
    this.payoffTools = el('div', { class: 'card-tools' });
    this.payoffCard = el('div', { class: 'card' },
      el('div', { class: 'card-head' },
        el('div', {},
          el('h3', {}, 'Profit and loss'),
          el('p', {}, 'The kinked line is P&L at expiry. The smooth curves are the position revalued at '
            + 'earlier dates — which is what the position actually does, and often a very different '
            + 'shape from the expiry diagram.')),
        this.payoffTools),
      this.payoffHost, this.payoffLegend);

    this.statsBox = el('div');
    this.greekHost = el('div');
    this.greekLegend = el('div');
    this.greekCard = el('div', { class: 'card' },
      el('div', { class: 'card-head' },
        el('div', {},
          el('h3', {}, 'Aggregate Greeks across spot'),
          el('p', {}, 'Position Greeks summed over legs, at the current tenor. Where delta crosses zero '
            + 'the position is momentarily hedged; gamma tells you how long that lasts.')),
        el('div', { class: 'card-tools' })),
      this.greekHost, this.greekLegend);

    this.host.append(
      this.notice,
      el('div', { class: 'grid grid-2' }, builderCard, el('div', { class: 'card' },
        el('div', { class: 'card-head' }, el('div', {}, el('h3', {}, 'Position summary'))),
        this.statsBox)),
      this.payoffCard,
      this.greekCard,
    );

    this.payoff = new LineChart(this.payoffHost, {
      height: 340, legendHost: this.payoffLegend, xLabel: 'spot at valuation', yLabel: 'profit / loss',
    });
    this.greeks = new LineChart(this.greekHost, {
      height: 300, legendHost: this.greekLegend, xLabel: 'spot', yLabel: 'position greek',
    });
    this.payoffTable = withTableToggle(this.payoff, this.payoffTools);
    this.payoffCard.append(this.payoffTable.holder);
  }

  async refresh() {
    this.dirty = false;
    const sig = signature();
    const T = Math.max(state.T, 1 / 365);
    const body = {
      S: state.S, T, r: state.r, q: state.q, sigma: state.sigma,
      horizons: [T, T * 0.6, T * 0.25, 0],
    };
    if (this.legs) body.legs = this.legs;
    else body.preset = this.preset;

    this.payoff.setStale(true);
    try {
      const res = await api.strategy(body);
      if (sig !== signature()) return;
      this.data = res;
      this.legs = res.legs;
      this._renderLegs();
      this._renderStats();
      this._renderPayoff();
      this._renderGreeks();
    } finally {
      this.payoff.setStale(false);
    }
  }

  _recalc() {
    this.refresh().catch((e) => this.ctx.reportError(e, 'strategy'));
  }

  _renderLegs() {
    const legs = this.legs || [];
    const rows = legs.map((leg, i) => {
      const kindSel = el('select', {
        onchange: (e) => { this.legs[i].kind = e.target.value; this._recalc(); },
      }, ...['call', 'put', 'underlying'].map((k) => el('option', { value: k, selected: k === leg.kind }, k)));

      const qty = el('input', {
        type: 'number', value: String(leg.quantity), step: '1', style: { width: '68px' },
        onchange: (e) => { this.legs[i].quantity = Number(e.target.value) || 0; this._recalc(); },
      });
      const strike = el('input', {
        type: 'number', value: String(leg.strike || 0), step: '0.5', style: { width: '84px' },
        disabled: leg.kind === 'underlying',
        onchange: (e) => { this.legs[i].strike = Number(e.target.value) || 0; this._recalc(); },
      });
      const vol = el('input', {
        type: 'number', value: (leg.sigma * 100).toFixed(2), step: '0.5', style: { width: '76px' },
        disabled: leg.kind === 'underlying',
        onchange: (e) => { this.legs[i].sigma = (Number(e.target.value) || 0) / 100; this._recalc(); },
      });
      const del = el('button', {
        class: 'btn btn-ghost btn-sm', type: 'button',
        onclick: () => { this.legs.splice(i, 1); if (!this.legs.length) this.legs = null; this._recalc(); },
      }, '×');

      return el('tr', {},
        el('td', {}, kindSel),
        el('td', {}, qty),
        el('td', {}, strike),
        el('td', {}, vol),
        el('td', { class: 'num' }, leg.premium !== null && leg.premium !== undefined ? fmtSig(leg.premium, 4) : el('span', { class: 'dim' }, 'theoretical')),
        el('td', {}, del));
    });

    clear(this.legsBox).append(
      el('div', { class: 'table-wrap', style: { marginTop: '12px' } },
        el('table', { class: 'data legs-table' },
          el('thead', {}, el('tr', {},
            el('th', {}, 'Instrument'), el('th', {}, 'Qty'), el('th', {}, 'Strike'),
            el('th', {}, 'Vol %'), el('th', {}, 'Entry'), el('th', {}, ''))),
          el('tbody', {}, rows))));
  }

  _renderStats() {
    const d = this.data;
    const g = d.net_greeks;
    const units = d.greek_units || {};

    const tiles = [
      { label: d.position === 'credit' ? 'Net credit' : 'Net debit', value: fmtMoney(Math.abs(d.entry_cost), 4),
        unit: d.position === 'credit' ? 'received to open' : 'paid to open' },
      { label: 'Max profit', value: d.max_profit === null ? 'unbounded' : fmtMoney(d.max_profit, 2),
        unit: d.max_profit === null ? 'a naked long tail' : 'at expiry' },
      { label: 'Max loss', value: d.max_loss === null ? 'unbounded' : fmtMoney(d.max_loss, 2),
        unit: d.max_loss === null ? 'a naked short tail' : 'at expiry' },
      { label: 'Breakevens', value: d.breakevens.length ? d.breakevens.map((b) => fmtSig(b, 5)).join('  ·  ') : 'none',
        unit: 'expiry P&L crosses zero' },
    ];

    clear(this.statsBox).append(
      el('div', { class: 'tiles' }, ...tiles.map((t) => el('div', { class: 'tile' },
        el('div', { class: 'tile-label' }, t.label),
        el('div', { class: 'tile-value' }, t.value),
        el('div', { class: 'tile-unit' }, t.unit)))),
      el('div', { class: 'panel-title', style: { marginTop: '14px' } }, 'Net position Greeks at spot'),
      el('div', { class: 'tiles' }, ...Object.entries(g).map(([name, v]) => el('div', { class: `tile ${v < 0 ? 'is-neg' : v > 0 ? 'is-pos' : ''}` },
        el('div', { class: 'tile-label' }, name),
        el('div', { class: 'tile-value' }, fmtSig(v, 4)),
        el('div', { class: 'tile-unit' }, units[name] || '')))),
    );

    clear(this.notice);
    if (d.max_loss === null) {
      this.notice.append(el('div', { class: 'notice warn' }, el('div', {},
        el('strong', {}, 'Unbounded loss. '),
        'This position has a naked short tail, so the loss is not capped by the grid — the payoff '
        + 'keeps going past the plotted range. The chart shows a window, not the whole risk.')));
    }
  }

  _renderPayoff() {
    const d = this.data;
    const curves = d.curves;
    const series = curves.map((c, i) => ({
      key: `c${i}`,
      label: c.label,
      shortLabel: c.days_out === 0 ? 'now' : c.label,
      values: c.pnl,
      color: seriesColor(i),
      area: i === curves.length - 1,
      areaPos: token('--series-3'),
      areaNeg: token('--critical'),
      width: i === curves.length - 1 ? 2.4 : 2,
    }));

    const markers = [
      { x: d.spot0, label: `spot ${fmtSig(d.spot0, 5)}`, color: token('--text-muted') },
      ...d.breakevens.map((b) => ({ x: b, label: `BE ${fmtSig(b, 5)}`, color: token('--border-strong') })),
    ];

    this.payoff.setData({
      x: d.spots,
      series,
      yZero: true,
      markers,
      xName: 'spot',
      xTitle: (v) => `spot ${fmtSig(v, 5)}`,
      valueFmt: (v) => fmtMoney(v, 3),
      yFmt: (v) => fmtMoney(v, 0),
    });
    this.payoffTable.refresh();
  }

  _renderGreeks() {
    const d = this.data;
    // Five Greeks is exactly the palette width; nothing is cycled.
    const names = Object.keys(d.greeks);
    this.greeks.setData({
      x: d.spots,
      series: names.map((name, i) => ({
        key: name,
        label: `${name} (${(d.greek_units || {})[name] || ''})`,
        shortLabel: name,
        values: d.greeks[name],
        color: seriesColor(i),
      })),
      yZero: true,
      markers: [{ x: d.spot0, label: `spot ${fmtSig(d.spot0, 5)}`, color: token('--text-muted') }],
      xName: 'spot',
      xTitle: (v) => `spot ${fmtSig(v, 5)}`,
      valueFmt: (v) => fmtSig(v, 5),
    });
  }

  redraw() {
    if (this.payoff) this.payoff.draw();
    if (this.greeks) this.greeks.draw();
  }
}

File: optlib/strategy.py
"""Multi-leg option strategy valuation: payoff, P&L and aggregate Greeks.

A leg is a signed quantity of a call, a put, or the underlying. The strategy is
evaluated two ways, and the distinction is the whole point of the view that uses
this module:

* **Payoff at expiry** -- intrinsic value only, the textbook kinked diagram.
* **Mark-to-market P&L** -- the position revalued at an arbitrary future date
  and spot with Black-Scholes, which is what the position actually does. A short
  strangle looks safe at expiry and terrifying a week in; only the second curve
  shows that.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np

from . import blackscholes as bs

LEG_KINDS = ("call", "put", "underlying")
AGGREGATE_GREEKS = ("delta", "gamma", "vega", "theta", "rho")


@dataclass
class Leg:
    kind: str          # "call" | "put" | "underlying"
    quantity: float    # signed; +1 long one contract-share, -1 short
    strike: float = 0.0
    sigma: float = 0.25
    premium: float | None = None   # entry price; theoretical if omitted
    multiplier: float = 1.0

    def validate(self) -> None:
        if self.kind not in LEG_KINDS:
            raise ValueError(f"leg kind must be one of {LEG_KINDS}, got {self.kind!r}")
        if self.kind != "underlying" and not (self.strike > 0):
            raise ValueError(f"{self.kind} leg needs a positive strike")
        if not np.isfinite(self.quantity):
            raise ValueError("leg quantity must be finite")


def _leg_value(leg: Leg, S, T: float, r: float, q: float):
    """Present value of one unit of the leg at spot S with T years remaining."""
    S = np.asarray(S, dtype=float)
    if leg.kind == "underlying":
        return S
    if T <= 0.0:
        return np.maximum(S - leg.strike, 0.0) if leg.kind == "call" else np.maximum(leg.strike - S, 0.0)
    return np.asarray(bs.price(S, leg.strike, T, r, q, leg.sigma, leg.kind == "call"), dtype=float)


def _leg_greek(leg: Leg, name: str, S, T: float, r: float, q: float):
    S = np.asarray(S, dtype=float)
    if leg.kind == "underlying":
        # The underlying has unit delta and no other sensitivity.
        return np.ones_like(S) if name == "delta" else np.zeros_like(S)
    if T <= 0.0:
        if name == "delta":
            itm = (S > leg.strike) if leg.kind == "call" else (S < leg.strike)
            return np.where(itm, 1.0 if leg.kind == "call" else -1.0, 0.0)
        return np.zeros_like(S)
    return np.asarray(bs.evaluate(name, S, leg.strike, T, r, q, leg.sigma, leg.kind == "call"), dtype=float)


def entry_cost(legs: list[Leg], S0: float, T0: float, r: float, q: float) -> float:
    """Net debit (positive) or credit (negative) to open the position."""
    total = 0.0
    for leg in legs:
        prem = leg.premium
        if prem is None:
            prem = float(np.asarray(_leg_value(leg, S0, T0, r, q)).reshape(()))
        total += leg.quantity * leg.multiplier * float(prem)
    return total


def evaluate(
    legs: list[Leg],
    S0: float,
    T0: float,
    r: float,
    q: float,
    spots=None,
    horizons=None,
    n_spots: int = 161,
    spot_span: float = 0.45,
) -> dict:
    """Payoff, mark-to-market P&L curves and aggregate Greeks for a strategy.

    `horizons` are years remaining at each revaluation date, largest first; 0.0 is
    expiry. Defaults walk from today to expiry so time decay is visible.
    """
    for leg in legs:
        leg.validate()
    if not legs:
        raise ValueError("a strategy needs at least one leg")

    if spots is None:
        spots = np.linspace(S0 * (1.0 - spot_span), S0 * (1.0 + spot_span), n_spots)
    spots = np.asarray(spots, dtype=float)

    if horizons is None:
        horizons = [T0, T0 * 0.5, T0 * 0.25, 0.0]
    horizons = [float(h) for h in horizons]

    cost = entry_cost(legs, S0, T0, r, q)

    curves = []
    for T in horizons:
        value = np.zeros_like(spots)
        for leg in legs:
            value += leg.quantity * leg.multiplier * _leg_value(leg, spots, T, r, q)
        curves.append({
            "T": T,
            "days_out": round((T0 - T) * 365.0, 1),
            "label": "At expiry" if T <= 0 else f"{(T0 - T) * 365.0:.0f}d from now",
            "value": value.tolist(),
            "pnl": (value - cost).tolist(),
        })

    greeks: dict[str, list[float]] = {}
    for name in AGGREGATE_GREEKS:
        agg = np.zeros_like(spots)
        for leg in legs:
            agg += leg.quantity * leg.multiplier * _leg_greek(leg, name, spots, T0, r, q)
        scale, _ = bs.GREEK_DISPLAY[name]
        greeks[name] = (agg * scale).tolist()

    payoff = np.asarray(curves[-1]["pnl"], dtype=float)
    # Breakeven spots: sign changes of the expiry P&L, refined by linear interpolation.
    sign = np.sign(payoff)
    idx = np.flatnonzero(sign[:-1] * sign[1:] < 0)
    breakevens = [
        float(spots[i] + (spots[i + 1] - spots[i]) * (0.0 - payoff[i]) / (payoff[i + 1] - payoff[i]))
        for i in idx
    ]

    # An unbounded tail is reported as such rather than as the grid edge, which
    # would silently understate the risk of a naked short leg.
    edge_slope_lo = payoff[1] - payoff[0]
    edge_slope_hi = payoff[-1] - payoff[-2]
    return {
        "spots": spots.tolist(),
        "spot0": S0,
        "entry_cost": cost,
        "position": "debit" if cost > 0 else ("credit" if cost < 0 else "flat"),
        "curves": curves,
        "greeks": greeks,
        "breakevens": breakevens,
        "max_profit": None if edge_slope_hi > 1e-9 or edge_slope_lo < -1e-9 else float(payoff.max()),
        "max_loss": None if edge_slope_hi < -1e-9 or edge_slope_lo > 1e-9 else float(payoff.min()),
        "payoff_at_spot0": float(np.interp(S0, spots, payoff)),
        "net_greeks": {name: float(np.interp(S0, spots, np.asarray(vals))) for name, vals in greeks.items()},
    }


# Preset strategies, parameterized off spot so they stay sensible at any level.
def presets(S0: float, T: float, sigma: float) -> dict[str, list[dict]]:
    lo, hi = round(S0 * 0.95, 2), round(S0 * 1.05, 2)
    far_lo, far_hi = round(S0 * 0.9, 2), round(S0 * 1.1, 2)
    atm = round(S0, 2)
    return {
        "long_call": [{"kind": "call", "quantity": 1, "strike": atm, "sigma": sigma}],
        "long_put": [{"kind": "put", "quantity": 1, "strike": atm, "sigma": sigma}],
        "covered_call": [
            {"kind": "underlying", "quantity": 1},
            {"kind": "call", "quantity": -1, "strike": hi, "sigma": sigma},
        ],
        "bull_call_spread": [
            {"kind": "call", "quantity": 1, "strike": atm, "sigma": sigma},
            {"kind": "call", "quantity": -1, "strike": hi, "sigma": sigma},
        ],
        "bear_put_spread": [
            {"kind": "put", "quantity": 1, "strike": atm, "sigma": sigma},
            {"kind": "put", "quantity": -1, "strike": lo, "sigma": sigma},
        ],
        "straddle": [
            {"kind": "call", "quantity": 1, "strike": atm, "sigma": sigma},
            {"kind": "put", "quantity": 1, "strike": atm, "sigma": sigma},
        ],
        "short_strangle": [
            {"kind": "call", "quantity": -1, "strike": hi, "sigma": sigma},
            {"kind": "put", "quantity": -1, "strike": lo, "sigma": sigma},
        ],
        "iron_condor": [
            {"kind": "put", "quantity": 1, "strike": far_lo, "sigma": sigma},
            {"kind": "put", "quantity": -1, "strike": lo, "sigma": sigma},
            {"kind": "call", "quantity": -1, "strike": hi, "sigma": sigma},
            {"kind": "call", "quantity": 1, "strike": far_hi, "sigma": sigma},
        ],
        "butterfly": [
            {"kind": "call", "quantity": 1, "strike": lo, "sigma": sigma},
            {"kind": "call", "quantity": -2, "strike": atm, "sigma": sigma},
            {"kind": "call", "quantity": 1, "strike": hi, "sigma": sigma},
        ],
        "calendar_ratio": [
            {"kind": "call", "quantity": -1, "strike": atm, "sigma": sigma},
            {"kind": "call", "quantity": 2, "strike": hi, "sigma": sigma},
        ],
    }
